### ISSUE TITLE
Webhook notify retry too many times

### ISSUE THREAD

**Issue by bigHave** [prometheus/alertmanager#2286]
**What did you do?**

In our prod environment, we use alertmanger to manage alerts fired from prometheus, and deliver alerts to another notify service by webhook receiver, as we use our own notify service to manage alert status and record alert history and so on.  Once there our notify service is down for about 10 minutes when alertmanager happen to deliver a firing alert to it. 
**What did you expect to see?**

alertmanager try to deliver that alert a few times and then abort.
**What did you see instead? Under which circumstances?**
Then I find in our record system that alertmanager has tried to deliver the very same alert so many times, like 200 times in 10 minutes.

So I did a experiment to see how webhook retry works.

**Environment**
I build alertmanager from souce (master branch) and run with following configuration(no cluster peer, just single instane), log level set to debug.

* Alertmanager configuration file:
```
global:
  resolve_timeout: 5m

route:
  group_by: ['alertname']
  group_wait: 10s
  group_interval: 10s
  repeat_interval: 1h
  receiver: 'web.hook'
receivers:
- name: 'web.hook'
  webhook_configs:
  - url: 'http://127.0.0.1:10086/alertmanager_test'
inhibit_rules:
  - source_match:
      severity: 'critical'
    target_match:
      severity: 'warning'
    equal: ['alertname', 'dev', 'instance']

```
where 127.0.0.1:10086/alertmanager_test was a simple http server that always response http status 500.

some of logs went as follows
* Logs:
```
level=debug ts=2020-06-01T12:20:12.936Z caller=dispatch.go:138 component=dispatcher msg="Received alert" alert=server down![1d7ec64][active]
level=debug ts=2020-06-01T12:20:12.936Z caller=dispatch.go:473 component=dispatcher aggrGroup="{}:{alertname=\"server down!\"}" msg=flushing alerts=[server down![1d7ec64][active]]
level=debug ts=2020-06-01T12:20:12.937Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=1 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:13.417Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=2 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:14.338Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=3 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:15.292Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=4 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:17.274Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=5 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:20.953Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=6 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=error ts=2020-06-01T12:20:22.941Z caller=notify.go:372 component=dispatcher msg="Error on notify" err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test" context_err="context deadline exceeded"
level=error ts=2020-06-01T12:20:22.941Z caller=dispatch.go:309 component=dispatcher msg="Notify for alerts failed" num_alerts=1 err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:22.941Z caller=dispatch.go:473 component=dispatcher aggrGroup="{}:{alertname=\"server down!\"}" msg=flushing alerts=[server down![1d7ec64][active]]
level=debug ts=2020-06-01T12:20:22.942Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=1 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:23.424Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=2 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:23.897Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=3 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:25.437Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=4 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:27.635Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=5 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:29.339Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=6 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:31.517Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=7 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=error ts=2020-06-01T12:20:32.942Z caller=notify.go:372 component=dispatcher msg="Error on notify" err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test" context_err="context deadline exceeded"
level=error ts=2020-06-01T12:20:32.942Z caller=dispatch.go:309 component=dispatcher msg="Notify for alerts failed" num_alerts=1 err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:32.942Z caller=dispatch.go:473 component=dispatcher aggrGroup="{}:{alertname=\"server down!\"}" msg=flushing alerts=[server down![1d7ec64][active]]
level=debug ts=2020-06-01T12:20:32.942Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=1 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:33.328Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=2 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:34.039Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=3 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:34.703Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=4 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:35.901Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=5 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=debug ts=2020-06-01T12:20:38.792Z caller=notify.go:668 component=dispatcher msg="Notify attempt failed" attempt=6 integration=webhook receiver=web.hook retry=true err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test"
level=error ts=2020-06-01T12:20:42.947Z caller=notify.go:372 component=dispatcher msg="Error on notify" err="unexpected status code 500: http://127.0.0.1:10086/alertmanager_test" context_err="context deadline exceeded"
```

after that,  I checked code related, and I see the webhook retry logic:
1. alert group will be flushed every group_interval time.
2. during the flush process, alerts will be delivered under a exponential backoff retry mechanism, which use a default setting that will retry about 28 times during  15 minutes(max elapsed time), unless this round of flush finished.

this looks not reasonable to me, according to our daily usage of alertmanager.

I created a PR related: # #2288

**Comment by simonpasquier** [comment]
I agree that not checking that the backoff timer has stopped is an issue. But I would consider that Alertmanager should retry as long as the group context isn't canceled. So we should set `MaxElapsedTime` to 0 rather than the default value of 15m as returned by `backoff.NewExponentialBackOff()`.

**Comment by bigHave** [comment]
Thank you for your reply, but sorry I didn't get the point.  Could you please discuss more about why Alertmanager should retry as long as the group context isn't canceled? because the way I see it is quite opposite to your opinion. I would consider that alertmanager should retry less, and backoff timer should stop sooner, because next round of flush would retry the same alert anyway.

**Comment by simonpasquier** [comment]
Thanks to your report, I realize that with a `group_interval` value greater than 15m ([`backoff.DefaultMaxElapsedTime`](https://pkg.go.dev/github.com/cenkalti/backoff/v4?tab=doc#pkg-constants)), Alertmanager will retry without back-off after the 15 minutes have elapsed.

> I would consider that alertmanager should retry less, and backoff timer should stop sooner, because next round of flush would retry the same alert anyway.

IMHO, Alertmanager should retry as long as the group's context hasn't been canceled. Maybe we could adjust the back-off parameters to be less "aggressive" but remember that 10 seconds for `group_interval` is well below the [recommendations](https://prometheus.io/docs/alerting/configuration/#route).

**Comment by bigHave** [comment]
Agreed. Thanks for your opinion. Please consider PR #2291

**Comment by simonpasquier** [comment]
Closed by #2290.
